### DISCUSSION TITLE
Connection pool timeout error  in GCP while using Cron services.

### DISCUSSION THREAD

**Discussion by MURIUKI2092** [prisma/prisma#22260] [Q&A]
### Question

I am getting this error when I deploy my app on Google Cloud. Am using Nest Js Cron Services and locally, the services are running  and do as expected.

```
DEFAULT 2023-12-05T09:20:34.171885Z Invalid `prisma.company.findMany()` invocation:
DEFAULT 2023-12-05T09:20:34.171901Z Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)[39m
DEFAULT 2023-12-05T09:21:20.475104Z /app/prisma/prisma/pms_system_db/runtime/library.js:123
```
```
DEFAULT 2023-12-05T09:21:20.475243Z `)}var Bi=(e,t)=>r=>{let n="",i;if(Array.isArray(r)){let[o,...s]=r;n=o,i={values:mr(s||[]),__prismaRawParameters__:!0}}else switch(e){case"sqlite":case"mysql":{n=r.sql,i={values:mr(r.values),__prismaRawParameters__:!0};break}case"cockroachdb":case"postgresql":{n=r.text,i={values:mr(r.values),__prismaRawParameters__:!0};break}case"sqlserver":{n=Ya(r),i={values:mr(r.values),__prismaRawParameters__:!0};break}default:throw new Error(`The ${e} provider does not support ${t}`)}return i?.values?tl(`prisma.${t}(${n}, ${i.values})`):tl(`prisma.${t}(${n})`),{query:n,parameters:i}},rl={requestArgsToMiddlewareArgs(e){return[e.strings,...e.values]},middlewareArgsToRequestArgs(e){let[t,...r]=e;return new oe(t,r)}},nl={requestArgsToMiddlewareArgs(e){return[e]},middlewareArgsToRequestArgs(e){return e[0]}};function Ui(e){return function(r){let n,i=(o=e)=>{try{return o===void 0||o?.kind==="itx"?n??(n=il(r(o))):il(r(o))}catch(s){return Promise.reject(s)}};return{then(o,s){return i().then(o,s)},catch(o){return i().catch(o)},finally(o){return i().finally(o)},requestTransaction(o){let s=i(o);return s.requestTransaction?s.requestTransaction(o):s},[Symbol.toStringTag]:"PrismaPromise"}}}function il(e){return typeof e.then=="function"?e:Promise.resolve(e)}var ol={isEnabled(){return!1},getTraceParent(){return"00-10-10-00"},async createEngineSpan(){},getActiveContext(){},runInChildSpan(e,t){return t()}},Ki=class{isEnabled(){return this.getGlobalTracingHelper().isEnabled()}getTraceParent(t){return this.getGlobalTracingHelper().getTraceParent(t)}createEngineSpan(t){return this.getGlobalTracingHelper().createEngineSpan(t)}getActiveContext(){return this.getGlobalTracingHelper().getActiveContext()}runInChildSpan(t,r){return this.getGlobalTracingHelper().runInChildSpan(t,r)}getGlobalTracingHelper(){return globalThis.PRISMA_INSTRUMENTATION?.helper??ol}};function sl(e){return e.includes("tracing")?new Ki:ol}function al(e,t=()=>{}){let r,n=new Promise(i=>r=i);return{then(i){return--e===0&&r(t()),i?.(n)}}}function ll(e){return typeof e=="string"?e:e.reduce((t,r)=>{let n=typeof r=="string"?r:r.level;return n==="query"?t:t&&(r==="info"||t==="info")?"info":n},void 0)}var Fd=["$connect","$disconnect","$on","$transaction","$use","$extends"],ul=Fd;var hn=class{constructor(){this._middlewares=[]}use(t){this._middlewares.push(t)}get(t){return this._middlewares[t]}has(t){return!!this._middlewares[t]}length(){return this._middlewares.length}};var pl=S(ni());function xn(e){return typeof e.batchRequestIdx…
DEFAULT 2023-12-05T09:21:20.475274Z
DEFAULT 2023-12-05T09:21:20.475289Z PrismaClientKnownRequestError:
DEFAULT 2023-12-05T09:21:20.475294Z Invalid `prisma.lease.findFirst()` invocation:
DEFAULT 2023-12-05T09:21:20.475307Z Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)
DEFAULT 2023-12-05T09:21:20.475315Z at wn.handleRequestError (/app/prisma/prisma/pms_system_db/runtime/library.js:123:6730)
DEFAULT 2023-12-05T09:21:20.475321Z at wn.handleAndLogRequestError (/app/prisma/prisma/pms_system_db/runtime/library.js:123:6119)
DEFAULT 2023-12-05T09:21:20.475331Z at wn.request (/app/prisma/prisma/pms_system_db/runtime/library.js:123:5839)
DEFAULT 2023-12-05T09:21:20.475338Z at async l (/app/prisma/prisma/pms_system_db/runtime/library.js:128:9763)
DEFAULT 2023-12-05T09:21:20.475344Z at async /app/dist/Notifications/messages.repository.js:41:27 {
DEFAULT 2023-12-05T09:21:20.475350Z code: 'P2024',
DEFAULT 2023-12-05T09:21:20.475358Z clientVersion: '5.3.1',
DEFAULT 2023-12-05T09:21:20.475364Z meta: { connection_limit: 5, timeout: 10 }
DEFAULT 2023-12-05T09:21:20.475369Z }
DEFAULT 2023-12-05T09:21:20.475375Z Node.js v18.19.0
DEFAULT 2023-12-05T09:21:25.174266Z 2023/12/05 09:21:25 Reading data from local connection on read: connection reset by peer
WARNING 2023-12-05T09:21:25.184095Z Container terminated on signal 11.
```
### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @MURIUKI2092 👋 

Thank you for raising this question.

 Have you tried increasing the [`connection_limit`](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#recommended-connection-pool-size) parameter in your database connection string to allow more concurrent connections to the database?

**Reply by MURIUKI2092** [reply]
Hi @ludralph 👋.
This is my local db string,
### postgresql://james:james@localhost:5432/system_db?connection_limit=5&pool_timeout=0.

On GCP, this is the structure for the db_uri:
### postgres://username:pass@localhost/client_db?host=/cloudsql/project:region:db

However, when I add that last bit:
### postgres://username:pass@localhost/client_db?host=/cloudsql/project:region:db?connection_limit=5&pool_timeout=0

it breaks on build

**Reply by ludralph** [reply]
Hi @MURIUKI2092 

You've correctly included `?host=/cloudsql/project:region:db` in the connection string but accidentally placed `? connection_limit=5&pool_timeout=0` after the host specification.  Simply move these parameters after the database name, like this:

```
postgres://username:pass@localhost/client_db?host=/cloudsql/project:region:db&connection_limit=5&pool_timeout=0
```

**Reply by MURIUKI2092** [reply]
Hi @ludralph  
Even after adding the string am still getting this error in the GCP console

Here is the error
`DEFAULT 2023-12-14T08:20:14.548897Z [31m[Nest] 1 - [39m12/14/2023, 8:20:14 AM [31m ERROR[39m [38;5;3m[Scheduler] [39m[31mPrismaClientKnownRequestError:
DEFAULT 2023-12-14T08:20:14.548926Z Invalid `prisma.mpesaTransactions.findMany()` invocation:
DEFAULT 2023-12-14T08:20:14.548940Z Can't reach database server at "database location"
DEFAULT 2023-12-14T08:20:14.548951Z Please make sure your database server is running at `database location`.[39m

It's passing the build now but breaking upon query execution.

**Reply by ludralph** [reply]
Hi @MURIUKI2092, can you access your database using a database GUI tool like [tableplus](https://tableplus.com/) or [DataGrip](https://www.jetbrains.com/datagrip)? Cloud SQL instances are configured to block all traffic to the public IP unless you specifically allow an IP using [Authorized Networks](https://cloud.google.com/sql/docs/postgres/authorize-networks). You may need to configure that as well.

**Reply by MURIUKI2092** [reply]
Hi @ludralph . I am able to access all my databases locally; that is when I add my ip address to the GCP connection lists. The problem is that I have done an introspect to the db giving me a timeout. Once the db  connection is opened to it's main app, then the Prisma Client throws the error described above.

**Reply by AxelDeutsch** [reply]
@MURIUKI2092 Did you solve the issue? If yes, how? I have the same problem, Thx in advance!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
